nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/PublishMQTT.java

```diff
@@ -0,0 +1,237 @@
+            .name("Retain Message")
+            .description("Whether or not the retain flag should be set on the MQTT message.")
+            .required(true)
+            .expressionLanguageSupported(true)
+            .addValidator(RETAIN_VALIDATOR)
+            .build();
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("FlowFiles that are sent successfully to the destination are transferred to this relationship.")
+            .build();
+    public static final Relationship REL_FAILURE = new Relationship.Builder()
+            .name("failure")
+            .description("FlowFiles that failed to send to the destination are transferred to this relationship.")
+            .build();
+
+    private static final List<PropertyDescriptor> descriptors;
+    private static final Set<Relationship> relationships;
+
+    static {
+        final List<PropertyDescriptor> innerDescriptorsList = getAbstractPropertyDescriptors();
+        innerDescriptorsList.add(PROP_TOPIC);
+        innerDescriptorsList.add(PROP_QOS);
+        innerDescriptorsList.add(PROP_RETAIN);
+        descriptors = Collections.unmodifiableList(innerDescriptorsList);
+
+        final Set<Relationship> innerRelationshipsSet = new HashSet<>();
+        innerRelationshipsSet.add(REL_SUCCESS);
+        innerRelationshipsSet.add(REL_FAILURE);
+        relationships = Collections.unmodifiableSet(innerRelationshipsSet);
+    }
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        logger = getLogger();
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        buildClient(context);
+    }
+
+    @OnStopped
+    public void onStop(final ProcessContext context) {
+        try {
+            synchronized (mqttClientConnectLock) {
+                if (mqttClient != null && mqttClient.isConnected()) {
+                    mqttClient.disconnect();
+                    logger.info("Disconnected the MQTT client.");
+                }
+            }
+        } catch(MqttException me) {
+            logger.error("Failed when disconnecting the MQTT client.", me);
+        }
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowfile = session.get();
+        if (flowfile == null) {
+            return;
+        }
+
+        if(mqttClient == null || !mqttClient.isConnected()){
+            logger.info("Was disconnected from client or was never connected, attempting to connect.");
+            synchronized (mqttClientConnectLock){
```

Comment:
Actually I should probably a introduce read/write lock for PublishMQTT on the MqttClient instead of just a basic lock